# 第二回レポート課題「条件数condの計算」

## 学籍番号：08B22182　　氏名：平山聖輝

課題内容：以下の条件数cond(A)を求める．行列Aは以下

$$
A =  \left(\begin{array}{cc}  
        4 & 2 \\ 
        1 & 3 
     \end{array}\right)
$$

In [4]:
import numpy as np
from numpy import linalg

In [10]:
A = np.array([[4,2], [1,3]])
print(A)

[[4 2]
 [1 3]]


まず，条件数(cond)について定義を確認すると行列Aが正則な時，講義の内容から以下のような式で表されることが分かる．（Aが正則でなければ条件数は単に $ \infty $ になる）

$$
cond(A) = ||A|| \cdot ||A^{-1}||
$$

ここで行列のノルムが登場するが，今回の課題では1ノルムと2ノルムの二つの場合で条件数を求める．

### 1ノルムのとき

行列の1ノルムは各列の成分の絶対和のうち最大のものである．

$$
||A||_1 = \max_j \sum^{n}_{i=1}{|a_{ij}|}
$$

よってまずはAの逆行列を求める必要がある．ここで逆行列についてだが，行列のサイズが大きくなると逆行列を求める計算コストが大きくなってしまうためなるべく避けなければならないことは知っておく必要がある．

In [34]:
Ainv = np.linalg.inv(A)
print(Ainv)

[[ 0.3 -0.2]
 [-0.1  0.4]]


上の $ A^{-1} $ の結果と，初めに表示した $ A $ の結果を見ると $ A $, $ A^{-1} $ の1ノルムはそれぞれ

\begin{align}
&||A|| = \max{[(4+1), (2+3)]} = 5 \\
&||A^{-1}|| = \max{[(0.3+0.1), (0.2+0.4)]} = 0.6
\end{align}

となることが分かるから1ノルムにおける条件数は
$$ cond(A)_1 = 5 \times 0.6 = 3 $$
ということが分かる．実際に1ノルムをそれぞれ計算すると

In [60]:
A1 = np.linalg.norm(A,1) # A1はAの1ノルム
Ainv1 = np.linalg.norm(Ainv,1) # Ainv1はA^-1の1ノルム
cond1 = A1 * Ainv1
print(str(A1) + "×" + str(Ainv1) + "=" + str(cond1))
print(np.linalg.cond(A,1))

5.0×0.6000000000000001=3.0000000000000004
3.0000000000000004


と両者が一致している様子が計算機上でも確認できた．（17桁目の値が少しおかしく，調べたところ小数を2進数で表すには分数の近似を行っていることが原因だった．10進法でも $ 3 / 10 $ が $ 0.333333... $であるように2進法では $ 1/10 $ が $ 0.00011001100... $となってしまうようだ）

### 2ノルムのとき

行列の2ノルムは条件数は $ A^{T}A $ の最大固有値と最小固有値の比になり，式で表すと

$$
cond(A) = \sqrt{\frac{\lambda_{max}(A^{T}A)}{\lambda_{min}(A^{T}A)}}
$$

となることは講義中でも紹介された．2ノルムの利点は $ A^{-1} $ を計算することなく，条件数を求められることにある．$ A $ の定義から $ A^{T}A $ は以下のように計算できる．

In [103]:
A = np.array([[4, 2],[1, 3]])
AT = np.array([[4, 1],[2, 3]])
ATA = AT @ A # 行列の積は" @ "で行う．" * "をすると要素同士を掛け合わせるだけになる
print(str(ATA))

[[17 11]
 [11 13]]


$$
A^{T}A = \left(\begin{array}{cc}
    4 & 1 \\
    2 & 3 
    \end{array}\right)
    \left(\begin{array}{cc}
    4 & 2 \\
    1 & 3 
    \end{array}\right) =
     \left(\begin{array}{cc}
    17 & 11 \\
    11 & 13 
    \end{array}\right)
$$

したがって $ A^{T}A $ の固有値は以下のようになる．

In [97]:
np.linalg.eig(ATA)

EigResult(eigenvalues=array([26.18033989,  3.81966011]), eigenvectors=array([[ 0.76775173, -0.64074744],
       [ 0.64074744,  0.76775173]]))

実際に固有多項式を解くと
$$ 
\begin{array}{|cc|}
    (\lambda - 17) & -11 \\
    -11 & (\lambda -13) 
\end{array} =
(\lambda-17)(\lambda-13)-121 = \lambda^2-30\lambda+100 =0
$$ $$
\Leftrightarrow \lambda = 15 \pm 5\sqrt{15} \approx 26.18033989, 3.8196011
$$


と確かにそれぞれの固有値が求まっていることが確かめられる．$ A $ が $ 2 \times 2 $ の行列であるから固有値は２つだけで，最小最大は明らかに求まる．$ cond(A) \ge 1 $ であることから両者の比を求めると

In [131]:
l,v = np.linalg.eig(ATA) # linalg.eigは97行目を見るとわかるように固有値と固有ベクトルの二つのデータを保管しているので固有値ベクトルの入っている配列をvでちゃんと定義してやる必要がある
print(np.sqrt(max(l)/min(l)))

2.618033988749895


と2ノルムにおける条件数が求まった．確認すると

In [124]:
print(np.linalg.cond(A,2))

2.618033988749895


と値が同じことが確認できる．
$$
cond(A)_2 = 2.618033988749895
$$